2023/5/one.hs

```diff
@@ -0,0 +1,66 @@
+import Control.FromSum(fromEither)
+import Data.Maybe(fromMaybe)
+import Data.Monoid(Last(..), getLast)
+import Text.Parsec(endBy, eof, many, many1, noneOf, parse, sepBy)
+import Text.Parsec.Char(char, digit, string)
+
+
+data Range = Range Integer Integer Integer
+
+
+translate :: Integer -> Range -> Maybe Integer
+translate val (Range dst src len) =
+    if src <= val && val < src + len
+    then Just $ dst + val - src
+    else Nothing
+
+
+data Mapping = Mapping [Range]
+
+
+transform :: Mapping -> Integer -> Integer
+transform (Mapping ranges) target =
+    fromMaybe (error "impossible") . getLast . mconcat . map Last $ ((Just target) : map (translate target) ranges)
```

Comment:
Using `Last` forces you to translate through all the ranges. Maybe use `First` and append `Just target` to the very end.